src/NuGet.Core/NuGet.Protocol/Resources/PackageMetadataResourceV3.cs

```diff
@@ -38,37 +55,240 @@ public class PackageMetadataResourceV3 : PackageMetadataResource
+                return Enumerable.Empty<PackageSearchMetadataRegistration>();
+            }
+
+            var results = new List<PackageSearchMetadataRegistration>();
+
+            foreach (var registrationPage in registrationIndex.Items)
+            {
+                if(registrationPage==null)
+                {
+                    throw new InvalidDataException(registrationUri.AbsoluteUri);
+                }
+
+                var lower = NuGetVersion.Parse(registrationPage.Lower);
+                var upper = NuGetVersion.Parse(registrationPage.Upper);
+
+                if (range.IsItemRangeRequired(lower, upper))
+                {
+                    if (registrationPage.Items == null)
+                    {
+                        var rangeUri = registrationPage.Url;
+                        var leafRegistrationPage = await GetRegistratioIndexPageAsync(_client, rangeUri, packageId, lower, upper, httpSourceCacheContext, log, token);
+
+                        if(registrationPage==null)
+                        {
+                            throw new InvalidDataException(registrationUri.AbsoluteUri);
+                        }
+
+                        ProcessRegistrationPage(leafRegistrationPage, results, range, includePrerelease,includeUnlisted, metadataCache);
+                    }
+                    else
+                    {
+                        ProcessRegistrationPage(registrationPage, results, range, includePrerelease, includeUnlisted, metadataCache);
+                    }
+                }
+            }
+
+            return results;
+        }
+
+        /// <summary>
+        /// Process RegistrationIndex and return list of RegistrationPages.
+        /// </summary>
+        /// <typeparam name="T">Generic type</typeparam>
+        /// <param name="httpSourceResult">HttpSourceResult type object which comes from HttpSource, used for get stream.</param>
+        /// <param name="token">Cancellation token.</param>
+        /// <returns></returns>
+        internal async Task<T> ProcessRegistrationIndexAsync<T>(HttpSourceResult httpSourceResult, CancellationToken token)
+        {
+            var jsonSerializer = JsonSerializer.Create(JsonExtensions.ObjectSerializationSettings);
+
+            using (var streamReader = new StreamReader(httpSourceResult.Stream))
+            using (var jsonReader = new JsonTextReader(streamReader))
             {
-                return ParseMetadata(metadata);
+                var registrationIndex = jsonSerializer
+                    .Deserialize<T>(jsonReader);
+
+                return await Task.FromResult(registrationIndex);
             }
-            return null;
         }
 
-        private PackageSearchMetadataRegistration ParseMetadata(JObject metadata)
+        /// <summary>
+        /// Query RegistrationIndex from nuget server for Consolidate UI. This implementation optimized for performance so instead of keeping gian JObject in memory we use strong types.
```

Comment:
```suggestion
        /// Query RegistrationIndex from nuget server for Consolidate UI. This implementation optimized for performance, so instead of keeping giant JObject in memory, we use strong types.
```

Reply:
Fixed.